core/cache/lxdprofilewatcher.go

```diff
@@ -0,0 +1,226 @@
+//        exist on the machine.
+func (w *MachineAppLXDProfileWatcher) unitChange(topic string, value interface{}) {
+	w.notifyWatcherBase.mu.Lock()
+	var notify bool
+	defer func(notify *bool) {
+		w.notifyWatcherBase.mu.Unlock()
+		if *notify {
+			logger.Tracef("notifying due to add/remove unit requires lxd profile change machine-%s", w.machineId)
+			w.notify()
+		}
+	}(&notify)
+
+	names, okString := value.([]string)
+	unit, okUnit := value.(*Unit)
+	switch {
+	case okString:
+		logger.Tracef("stop watching %q on machine-%s", names, w.machineId)
+		notify = w.removeUnit(names)
+	case okUnit:
+		switch {
+		case unit.details.MachineId == "" && !unit.details.Subordinate:
+			logger.Tracef("%s has no machineId and not a sub", unit.details.Name)
+			return
+		case unit.details.Subordinate:
+			principal, err := w.getUnit(unit.details.Principal)
+			if err != nil {
+				logger.Tracef("unit %s is subordinate, principal %s not found", unit.details.Name, unit.details.Principal)
+				return
+			}
+			if w.machineId != principal.details.MachineId {
+				logger.Tracef("watching unit changes on machine-%s not machine-%s", w.machineId, unit.details.MachineId)
+				return
+			}
+		case w.machineId != unit.details.MachineId:
+			logger.Tracef("watching unit changes on machine-%s not machine-%s", w.machineId, unit.details.MachineId)
+			return
+		}
+		logger.Tracef("start watching %q on machine-%s", unit.details.Name, w.machineId)
+		notify = w.addUnit(unit)
+	default:
+		logger.Errorf("programming error, value not of type *Unit or []string")
+	}
+	logger.Tracef("end of unit change %#v", w.applications)
+}
+
+func (w *MachineAppLXDProfileWatcher) addUnit(unit *Unit) bool {
+	_, ok := w.applications[unit.details.Application]
+	if !ok {
+		curl := unit.details.CharmURL
+		if curl == "" {
+			// this happens for new units to existing machines.
+			app, err := w.getApp(unit.details.Application)
+			if err != nil {
+				logger.Errorf("failed to get application %s for machine-%s", unit.details.Application, w.machineId)
+				return false
+			}
+			curl = app.details.CharmURL
+		}
+		ch, err := w.getCharm(curl)
+		if err != nil {
+			logger.Errorf("failed to get charm %q for %s on machine-%s", curl, unit.details.Name, w.machineId)
+			return false
+		}
+		info := appInfo{
+			charmURL: curl,
+			units:    set.NewStrings(unit.details.Name),
+		}
+		if !ch.details.LXDProfile.Empty() {
+			info.charmProfile = &ch.details.LXDProfile
+		}
+		w.applications[unit.details.Application] = info
+	} else {
+		w.applications[unit.details.Application].units.Add(unit.details.Name)
+	}
+	if w.applications[unit.details.Application].charmProfile != nil {
+		return true
+	}
+	return false
+}
+
+func (w *MachineAppLXDProfileWatcher) removeUnit(names []string) bool {
+	if len(names) != 2 {
+		logger.Errorf("programming error, 2 values not provided")
+		return false
+	}
+	unitName, appName := names[0], names[1]
+	_, ok := w.applications[appName]
```

Comment:
to reduce the verbosity of the following code, you could do:

```go
app, ok := w.applications[appName]
```

Then every time it mentions `w.applications[appName]` we can use `app`. That should reduce the noise of the code.

Reply:
done